docs/DifferentiableFunctions.md

```diff
@@ -0,0 +1,444 @@
+# Differentiable functions and differentiation APIs
+
+[Richard Wei], [Dan Zheng], [Marc Rasi], [Parker Schuh]
+
+Last updated: March 2019
+
+## Introduction
+
+Swift supports differentiable functions as part of the language. The
+`@differentiable` attribute appears in two locations in Swift syntax: as an
+annotation on function types and as an annotation on function declarations (and
+other similar declarations). This document explains the meaning of these
+annotations.
+
+## `@differentiable` function type attribute
+
+### Basics
+
+In Swift, function types can have attributes. When a function type is annotated
+with `@differentiable`, Swift guarantees that all values of that function type
+can be differentiated.
+
+`@differentiable` functions can be called like normal functions, or be passed to
+APIs that take `@differentiable` functions, like [`gradient(of:)`]. The binary
+representation of a `@differentiable` function is a special data structure
+containing the original function along with extra information required for
+computing its derivatives. Usage `@differentiable` functions are a part of
+Swift's type system. Most notably, they are used by differentiation APIs in the
+standard library. Here are some examples demonstrating differentiation APIs:
+
+```swift
+func square(_ x: Float) -> Float {
+    return x * x
+}
+let x: Float = 3.0
+
+// Free function examples.
+
+// Computes the gradient of `square` at `x`.
+print(gradient(at: x, in: square)) // 6.0
+// Computes the gradient of `square`, then applies it to `x`.
+print(gradient(of: square)(x)) // 6.0
+// Computes the value and gradient of `square` at `x`.
+print(valueWithGradient(at: x, in: square)) // (value: 9.0, gradient: 6.0)
```

Comment:
It may be nice to include examples of `pullback` and `valueWithPullback`.
Also, it may be nice to include examples of differentiation APIs for higher arity functions.